BEASTIE/extractHets.py

```diff
@@ -25,55 +39,98 @@ def count_all_het_sites(tmp,sample,vcfFilename,file_dir,outputFilename,chr_start
-                        cmd = "tabix " + vcfFilename + " "+chromN+":"+str(begin)+"-"+str(end)#
-                        #tabix /data/allenlab/scarlett/data/VCF/GSD/DNA_vcf/125249.vcf.recode.vcf.gz 1:10042358-10045556
-                        output=Pipe.run(cmd)
-                        if(len(output)==0):continue
-                        lines=output.split("\n")
-                        for line in lines:
-                            fields=line.split("\t")
-                            if(fields[6]!="PASS"): continue
-                            pos=fields[1]                         # column 9
-                            rs = fields[2]                        # column 10
-                            genotype = fields[9].split(':')[0]
-                            if(not isHeterozygous(str(genotype))):continue # go back to the begining of the loop
-                            transcriptCoord=transcript.mapToTranscript(int(pos))#
-                            total_biSNP += 1
-                            chr_pos=chromN+"_"+pos
-                            byGene[geneID].add(chr_pos)
-                            out_stream.write("\t".join([str(chrom),str(chromN),str(geneID),str(pos),str(transID),str(transcriptCoord),str(rs),str(genotype),"\n"]))
-            # write up the basic information
-            # all_set_list = []
-            # non_empty_count = 0
-            # non_empty_count_len = 0
-            # for each_key in byGene:
-            #     all_set_list.extend(list(byGene[each_key]))
-            #     if len(byGene[each_key])!= 0:
-            #         non_empty_count += 1
-            #         non_empty_count_len += len(byGene[each_key])
+
+                        region_str = f"{chromN}:{begin}-{end}"
+                        if not region_str in region_str_to_transcripts:
+                            region_str_to_transcripts[region_str] = []
+                        # else:
+                        #     for existing in region_str_to_transcripts[region_str]:
+                        #         print(f"existing transcript @ {region_str} {existing.getTranscriptId()} != {transcript.getTranscriptId()}")
+                        region_str_to_transcripts[region_str].append(transcript)
+            
+            print(f"{datetime.now()} found {len(region_str_to_transcripts)} regions to process")
+            CHUNK_SIZE = 1000
+            outputs = []
+            for x in chunk_iter(iter(region_str_to_transcripts.keys()), CHUNK_SIZE):
+                regions = ' '.join(x)
+                cmd = f"tabix --separate-regions {vcfFilename} {regions}"
+                output = Pipe.run(cmd)
+                if len(output) > 0:
+                    outputs.append(output)
+
+            print(f"{datetime.now()} got {len(outputs)} tabix call outputs")
+
+            out_stream = open(outputFile, "w")
+            out_stream.write("chr\tchrN\tgeneID\tpos\ttranscriptID\ttranscript_pos\tSNP_id\tgenotype\n")
+            for output in outputs:
```

Comment:
Instead of adding all the outputs to an array, then iterating through the array, I think it would be better to skip the `outputs` array entirely and process the results from `tabix` immediately. Maybe even pull the tabix output processing code into its own function.

Reply:
So my thought here is to process this in 3 explicit stages

1. compute regions to collect
2. collect regions in batches
3. process regions

I agree that it is an inefficiency and could be written as

1. compute regions
2. collected and process regions in batch

I don't feel super strongly about this, but would prefer to leave it as-is as I know it works.